### PR TITLE
Validation for RN headers

### PR DESCRIPTION
**Pull request by AradCarmi**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-4618)

## Description
Added a new validation to the **validate** command to verify that the release notes headers are in the correct format.

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,8 @@
 * Added support for fromversion/toversion in XSIAM content items (correlation rules, XSIAM dashboards, XSIAM reports and triggers).
 * Updated the supported python versions, as `>=3.8,<3.11`, as some of the dependencies are not supported on `3.11` yet.
 * Added **prepare-content** command which will prepare the pack or content item for the platform.
+* Added a new validation to the **validate** command to verify that the release notes header is in the correct format.
```
Comment (AradCarmi) [inline CHANGELOG.md:16]:
```suggestion
* Added a new validation to the **validate** command to verify that the release notes headers are in the correct format.
```

File: demisto_sdk/commands/common/errors.py [hunk 2]
```diff
@@ -17,17 +17,16 @@
     'BA101', 'BA106', 'BA108', 'BA109', 'BA110', 'BA111', 'BA112', 'BA113', 'BA116', 'BA119',
     'DS107',
     'GF102',
-    'IF100', 'IF106', 'IF113', 'IF115', 'IF116',
+    'IF100', 'IF106', 'IF115', 'IF116',
```
Comment (AradCarmi) [inline demisto_sdk/commands/common/errors.py:20]:
```suggestion
    'IF100', 'IF106', 'IF113', 'IF115', 'IF116',
```

File: demisto_sdk/commands/update_release_notes/update_rn.py [hunk 3]
```diff
@@ -798,7 +793,7 @@ def get_file_description(path, file_type) -> str:
         print_warning(f'Cannot get file description: "{path}" file does not exist')
         return ''
 
-    elif file_type in (FileType.PLAYBOOK, FileType.INTEGRATION, FileType.CORRELATION_RULE, FileType.MODELING_RULE):
+    elif file_type in (FileType.PLAYBOOK, FileType.INTEGRATION, FileType.CORRELATION_RULE):
```
Comment (AradCarmi) [inline demisto_sdk/commands/update_release_notes/update_rn.py:796]:
```suggestion
    elif file_type in (FileType.PLAYBOOK, FileType.INTEGRATION, FileType.CORRELATION_RULE, FileType.MODELING_RULE):
```

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 4]
```diff
@@ -36,6 +42,140 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+
+        content_type_dir_list = []
+        try:
+            for file in os.listdir(content_type_path):
+                if os.path.isdir(os.path.join(content_type_path, file)):
+                    file_path = os.path.join(content_type_path, file, f'{file}.yml')
+                else:
+                    file_path = os.path.join(content_type_path, file)
+                if file_path.endswith('.yml') | file_path.endswith('.json'):
+                    content_type_dir_list.append(file_path)
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:163]: can tools.get_files_in_dir() replace this?

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 5]
```diff
@@ -36,6 +42,140 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+    def validate_content_type_header(self, content_type: str) -> bool:
+        """
+            Validate that the release notes 1st headers (the content type) are a valid content entity.
+        Args:
+            content_type: (str) - The content type to validate.(e.g. Integrations, Playbooks, etc.)
+        Return:
+            True if the content type is valid, False otherwise.
+        """
+        # Get all the content type headers
+        rn_valid_headers = list(map(lambda rn_valid_header: rn_valid_header.lower(), RN_HEADER_BY_FILE_TYPE.values()))
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:132]: why not just `RN_HEADER_BY_FILE_TYPE.values()`?

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 6]
```diff
@@ -36,6 +42,140 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+
+        if content_type.lower() not in rn_valid_headers:
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:134]:
```suggestion
        if content_type not in rn_valid_headers:
```
needs to be case sensitive

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 7]
```diff
@@ -36,6 +42,140 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+                    file_path = os.path.join(content_type_path, file)
+                if file_path.endswith('.yml') | file_path.endswith('.json'):
+                    content_type_dir_list.append(file_path)
+        except FileNotFoundError:
+            error_message, error_code = Errors.release_notes_invalid_content_type_header(content_type=content_type,
+                                                                                         pack_name=self.pack_name)
+            if self.handle_error(error_message, error_code, self.release_notes_file_path):
+                is_valid = False
+
+        content_type_dir_list = list(map(lambda item: get_display_name(item), content_type_dir_list))
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:170]: it's ok - list comprehension is more readable IMO

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 8]
```diff
@@ -136,6 +276,27 @@ def validate_json_when_breaking_changes(self) -> bool:
                     is_valid = False
         return is_valid
 
+    @error_codes('RN113,RN114')
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:279]: we add a decorator of multiple error codes if all of them are checked in the method directly. this is not the case - then we can add the decorator for each function separately

File: demisto_sdk/commands/common/tests/test_files/rn_header_test_data [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+- Added x y z
+
+#### Scripts
+##### New: script-test
+- Added x y z
+
+#### Playbooks
+##### playbook-test.yml
+- Added x y z
+- Ignore the yml suffix, it is not a real playbook.
```
Comment (dantavori) [inline demisto_sdk/commands/common/tests/test_files/rn_header_test_data:12]: why does it pass then? (if it should - please add a test case for invalid RN headers)

**Review by dantavori — CHANGES_REQUESTED** [review]
nice, see notes

File: demisto_sdk/commands/common/hook_validations/release_notes.py (hunk 7 above)
Comment (AradCarmi) [reply demisto_sdk/commands/common/hook_validations/release_notes.py:170]: Agreed, but less efficient

File: demisto_sdk/commands/common/hook_validations/release_notes.py (hunk 8 above)
Comment (AradCarmi) [reply demisto_sdk/commands/common/hook_validations/release_notes.py:279]: The issue is that both of the functions need the pre-prepare extract and filter so how can we split it without duplicating some calls?

File: demisto_sdk/commands/common/tests/test_files/rn_header_test_data (hunk 9 above)
Comment (AradCarmi) [reply demisto_sdk/commands/common/tests/test_files/rn_header_test_data:12]: The `TestSuite` creates the content items this way, I just followed

File: demisto_sdk/commands/common/hook_validations/release_notes.py (hunk 8 above)
Comment (dantavori) [reply demisto_sdk/commands/common/hook_validations/release_notes.py:271]: shouldn't be an issue, please consult with @MLainer1 about it

**Review by dantavori — APPROVED** [review]
looks great, before merging let's make a short demo when you are available

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 10]
```diff
@@ -34,6 +40,134 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+            True if the content type is valid, False otherwise.
+        """
+        # Get all the content type headers
+        rn_valid_headers = RN_HEADER_BY_FILE_TYPE.values()
+        if content_type not in rn_valid_headers:
+            error_message, error_code = Errors.release_notes_invalid_content_type_header(content_type=content_type, pack_name=self.pack_name)
+            if self.handle_error(error_message, error_code, self.release_notes_file_path):
+                return False
+        return True
+
```
Comment (MLainer1) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:136]: Please add the '@error_codes' decorator here for the respective validation in this function.

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 11]
```diff
@@ -34,6 +40,134 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+            None.
+        """
+        for content_type, content_items in headers.items():
+            content_items = self.filter_nones(ls=content_items)
+            if content_type in RN_CONTENT_ENTITY_WITH_STARS and content_items:
+                headers[content_type] = [item.replace('*', '') for item in content_items]
+            if content_type.lower().replace(' ', '') in RN_CONTENT_ENTITY_WITH_STARS:
+                content_items = list(map(lambda x: x.replace('**', ''), content_items))
+            headers[content_type] = [item.replace('New:', '').strip() for item in content_items]
+
```
Comment (MLainer1) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:120]: Please add the '@error_codes' decorator here for the respective validation in the function below.

File: demisto_sdk/commands/common/hook_validations/release_notes.py [hunk 12]
```diff
@@ -34,6 +40,134 @@ def __init__(self, release_notes_file_path, modified_files=None, pack_name=None,
...
+            for content_type_section in content_type_sections_ls:
+                content_type_section = self.filter_nones(ls=content_type_section)
+                if content_type_section:
+                    header = content_type_section[0]
+                    if headers.get(content_type):
+                        headers[content_type].append(header)
+                    else:
+                        headers[content_type] = [header]
+        return headers
+
```
Comment (MLainer1) [inline demisto_sdk/commands/common/hook_validations/release_notes.py:84]: Please add the '@error_codes' decorator here for the respective validation in the function below.

**Comment by idovandijk** [comment]
Hey @AradCarmi can you please write what the fixed validation is looking for? And test with RN produced for various content items. Thanks

**Comment by AradCarmi** [comment]
@idovandijk 
The format of the release notes headers should be as follows (besides some exceptions):
#### <common-content-type>
##### <content item>
- Changes

The following content types have a special form:
#### <content type with star>
- Changes
- **<content item1>**
- **<content item2>**


content types with stars:
CONNECTION, INCIDENT_TYPE, REPUTATION, LAYOUT, INCIDENT_FIELD, INDICATOR_FIELD

Already test it with various content types but feel free to check it out.

**Review by dantavori — COMMENTED** [review]
great job! last notes

File: demisto_sdk/commands/common/constants.py [hunk 13]
```diff
@@ -212,6 +212,7 @@ class FileType(str, Enum):
     FileType.INCIDENT_FIELD.value: INCIDENT_FIELDS_DIR,
     FileType.INCIDENT_TYPE.value: INCIDENT_TYPES_DIR,
     FileType.INDICATOR_FIELD.value: INDICATOR_FIELDS_DIR,
+    FileType.INDICATOR_TYPE.value: INDICATOR_TYPES_DIR,
```
[the remark this answers is not in the document]
Comment (AradCarmi) [reply demisto_sdk/commands/common/constants.py]: Line 219:
`FileType.REPUTATION.value: INDICATOR_TYPES_DIR`
We need indicator type -> indicator type dir

**Pull request merged**